### PR TITLE
support multiple versions of packages

### PR DESCRIPTION
**Pull request by brainrake**
- support multiple sets of versions of packages
- make overlays for package sets
- use default overlay in module

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -5,14 +5,23 @@
     };
 
     # Cardano-node
-    cardano-node = {
+    "cardano-node-8.7.3" = {
       url = "github:intersectmbo/cardano-node?ref=8.7.3";
     };
-    cardano-configurations = {
+    "cardano-configurations-8.7.3" = {
```
Comment (aciceri) [inline flake.nix:11]: I just found out that [cardano-playground](https://github.com/input-output-hk/cardano-playground/) tags commits using the cardano-node versions. We can use that repository (always with `flake = false`) to fetch the configuration files.

File: flake.nix [hunk 2]
```diff
@@ -5,14 +5,23 @@
...
       # This version is compatible with cardano-node above and likely needs to be updated together.
       url = "github:input-output-hk/cardano-configurations/21249e0d5c68b4e8f3661b250aa8272a8785d678";
       flake = false;
     };
+    "cardano-node-8.1.1" = {
+      url = "github:intersectmbo/cardano-node?ref=8.7.3";
+    };
+    "cardano-configurations-8.1.1" = {
```
Comment (aciceri) [inline flake.nix:19]: As above.

File: packages/kupo.nix [hunk 3]
```diff
@@ -2,33 +2,38 @@
...
+      };
+    };
+    mkPackage = version:
+      pkgs.runCommandNoCC "kupo-${version}" {inherit version;} ''
         mkdir $out
         cd $out
-        ${pkgs.gnutar}/bin/tar -xvzf ${src}
+        ${pkgs.gnutar}/bin/tar -xvzf ${releases.${version}.${system}}
         chmod a+x $out/bin/kupo
       '';
```
Comment (aciceri) [inline packages/kupo.nix:30]: You can avoid this extra step by doing directly something like
```nix
pkgs.fetchzip {
  url = "https://github.com/CardanoSolutions/kupo/releases/download/v2.8/kupo-2.8.0-amd64-Linux.tar.gz";
  hash = "sha256-k6js0R0psyeHnM6q0e4slu4ESXm1FMiVRO2JUlsnlHY=";
  stripRoot = false;
  name = "kupo-2.8.0";
  postFetch = "chmod +x $out/bin/kupo";
}
```
It seems to produce the same output plus other possible useful files in `$out/share` like the manual and completions, if they don't really work we can remove them using the `postFetch` argument.

In general I believe that caching the uncompressed archive is better, I remember reading of a build system that was caching the compressed archives (zips or tars, can't remember) when at a certain point GitHub chose to change something in the compression algorithm they use to serve zips/tars, invalidating this way all the hashes.

File: packages/ogmios.nix
Comment (aciceri) [inline packages/ogmios.nix:1]: The same thing I wrote above for `kupo` is probably valid also here.

File: flake.nix (hunk 1 above)
Comment (brainrake) [reply flake.nix:11]: only since 8.7.3 :(

File: flake.nix (hunk 2 above)
Comment (brainrake) [reply flake.nix:19]: only since 8.7.3 :(

Comment (brainrake) [reply flake.nix:19]: but i messed up the copy paste above, will fix that too

Comment (aciceri) [reply flake.nix:19]: :cry:

**Review by aciceri — APPROVED** [review]
Looks perfect to me now, can we merge?

Edit: Oh no, I didn't notice that CI broke

**Review by aciceri — APPROVED** [review]
Can we merge it now?

File: flake.nix (hunk 1 above)
Comment (brainrake) [reply flake.nix:11]: let's address this in a separate PR. I added your comment to the issue: https://github.com/mlabs-haskell/cardano.nix/issues/52

File: packages/kupo.nix (hunk 3 above)
Comment (brainrake) [reply packages/kupo.nix]: fixed

File: packages/ogmios.nix
Comment (brainrake) [reply packages/ogmios.nix:1]: fixed

File: flake.nix (hunk 2 above)
Comment (brainrake) [reply flake.nix:19]: fixed cardano-node and cardano-configurations versions

**Comment by brainrake** [comment]
yep

**Pull request merged**